integrations/gcp/gcp_core/helpers/ratelimiter/base.py

```diff
@@ -0,0 +1,94 @@
+from abc import ABC, abstractmethod
+from typing import Any, Optional, TYPE_CHECKING
+from aiolimiter import AsyncLimiter
+from google.cloud.cloudquotas_v1 import CloudQuotasAsyncClient, GetQuotaInfoRequest  # type: ignore
+from loguru import logger
+from enum import Enum
+from port_ocean.context.ocean import ocean
+from gcp_core.cache import cache_coroutine_result  # type: ignore
+from collections.abc import MutableSequence
+
+
+_DEFAULT_RATE_LIMIT_TIME_PERIOD: float = 60.0
+_DEFAULT_RATE_LIMIT_QUOTA: int = int(
+    ocean.integration_config["search_all_resources_per_minute_quota"]
+)
+
+if TYPE_CHECKING:
+    from google.cloud.cloudquotas_v1 import DimensionsInfo
+
+
+class ContainerType(Enum):
+    PROJECT = "projects"
+    FOLDER = "folders"
+    ORGANIZATION = "organizations"
+
+
+class GCPResourceQuota(ABC):
+    quota_id: str | None = None
+    service: str | None = None
+    container_type: ContainerType = ContainerType.PROJECT
+    _default_quota: int = _DEFAULT_RATE_LIMIT_QUOTA
+
+    async def _request_quota_info(self, name: str) -> MutableSequence["DimensionsInfo"]:
+        async with AsyncLimiter(
+            max_rate=_DEFAULT_RATE_LIMIT_QUOTA,
+            time_period=_DEFAULT_RATE_LIMIT_TIME_PERIOD,
+        ):
+            async with CloudQuotasAsyncClient() as quotas_client:
+                request = GetQuotaInfoRequest(name=name)
+                response = await quotas_client.get_quota_info(request=request)
```

Comment:
does it requires extra permissions to perform this request? is it something we should add as part of our docs and TF?

Reply:
Yes it does require the Cloud API Quota API to be on for the project (container) and also some permission relating to the service being tracked allocated as such. Adding to the docs is vital

Reply:
We have to add this to the TF as well. Please specify which exact additional permissions are required here.

Reply:
what happens if there are no permissions to get it? how users upgrades to this permission?